### PR TITLE
./contributors/ documented + documented the procedure to update API Docs

### PR DESCRIPTION
**Pull request by ewenjo**
…dure to update API Docs - CA-107.

### CODE REVIEW

**Comment by tejas-hosamani** [comment]
[As a Developer, I would like to have the Contributors endpoints documented](https://trello.com/c/suX6Gtlk/107-as-a-developer-i-would-like-to-have-the-contributors-endpoints-documented)

File: package.json
Comment (ewenjo) [inline package.json]: Should be a dev dependency

File: src/app.js
Comment (ewenjo) [inline src/app.js]: Move all the code to a separate router file and include it in routes.js

**Review by ewenjo — COMMENTED** [review]
Just a few things before I learn more about swagger and do a proper review.

File: src/app.js
Comment (CodingCreate101) [reply src/app.js]: Moved Swagger code to routes.js.

Comment (tejas-hosamani) [reply src/app.js]: Alright, that should do it. Please review again.
More modular code, added a structure to maintain API versions as well.

**Comment by spiray** [comment]
@CodingCreate101 Thanks! I have a few questions/comments:
1. Does the APIs.json have to be in the root? Id want to keep it minimal if possible.
2. Can you put the docView folder in the api folder and follow the same structure with the separated routers and controllers? (Tests as well if possible).
3. The dev URL is `https://api-dev.coding.garden/`
4. The prod URL is `https://api.coding.garden/` I believe.
5. You can use `https://coding.garden/` as the website URL for now.

**Comment by tejas-hosamani** [comment]
> 1. Does the APIs.json have to be in the root? Id want to keep it minimal if possible.

Not necessarily. But where'd you want it otherwise? My reasoning is, I wanted to keep source files and Doc JSON file separate. 

> 2. Can you put the docView folder in the api folder and follow the same structure with the separated routers and controllers? (Tests as well if possible).

Yes, I could. In fact, I did put it there but then I realized, its not actually API. It's View. That's why I made a different folder altogether. Do you still think it should be in `api` folder? It has no controller. Also not sure how to add tests to it. I'd appreciate any suggestions. I 'll refer to other tests in the meantime.

> 3. The dev URL is `https://api-dev.coding.garden/`

Thanks. I 'll update that. But it doesn't seem to work.

> 4. The prod URL is `https://api.coding.garden/` I believe.

Could you please confirm. It's not up either.

> 5. You can use `https://coding.garden/` as the website URL for now.

Alright.

Thanks for the review.  @spiray  :)

**Comment by spiray** [comment]
1. I'm not sure :)
2. I think it can be in a docs folder in the api folder. @ewenjo Can you weigh in on this? 

`doc.routes.js`
```js
const router = require('express').Router();
const swaggerUi = require('swagger-ui-express');
const { sendVersionMarkup } = require('./doc.controller');
const openApiDocumentation = require('../../APIs.json');

const options = {
  // customCssUrl: '/custom.css' // If additional options ar needed.
};

router.use('/', swaggerUi.serve);

router.get('/', sendVersionMarkup);

router.get('/v1.0.0', swaggerUi.setup(openApiDocumentation, options));

module.exports = router;
```
`doc.controller.js`
```js
const path = require('path');

function sendVersionMarkup(req, res) {
  res.sendFile(path.join(__dirname, '/versions.html'));
}

module.exports = { sendVersionMarkup };
```
In terms of tests you can use supertest for testing status codes and then potentially use puppeteer for e2e tests. 

3. The dev url is working for me? What does it show for you? 
4.  The prod site isnt live yet as we haven't deployed anything to prod yet.

Thanks!

**Comment by tejas-hosamani** [comment]
1. Hahha. Alright, I know, it looks weird to just keep it in the open with `.gitignore`, `package.json`, etc,. I 'll put that in `/docs` folder. Time to get this file its own home.

2. My view is, `api` folder should have things that people actually call to consume raw data. Literally APIs. However, this file actually returns HTML content. React bundle, actually. That wouldn't quite fit in with API types. Let's see what ewenjo has in mind. :)
 - And hey, thanks a lot for those snippets. 
 - I'll try the tests.

3. It opens `https://api-dev.coding.garden/` with a nice message but actual APIs, for example, `https://api-dev.coding.garden/video`, returns an empty array. I'll update the URL though - since we actually have the URL that we do maintain.

4. I'll update the URL in docs. Hopefully, it will be functional soon. :) 

Thanks!

**Comment by ewenjo** [comment]
I would say the docView folder should be inside the API folder. It's cleaner and is has pretty much the same structure as the other files. 

https://api-dev.coding.garden/video is temporarily not working as intended, there is something wrong with the deployment which only CJ can fix. However, it is the dev url we will be using. Until the deployment is fixed, I'm hosted one that can be used for testing. https://dev.ewenjo.pro/

**Comment by spiray** [comment]
@CodingCreate101 The build failed because there is no localhost spun up. @ewenjo Do you have any idea how this should work? He would also run into issues with the dev deployment because the deployment would only take place dafter the PR is merged but the PR wont merge until the build passes. It could be I am missing something simple here.

**Comment by ewenjo** [comment]
I'm guessing it's failing because he hard-coded port 3000 into the test. Travis most likely changes the port when running the tests using the PORT env variable.

Or, because the test doesn't actually start the app, all the other tests require and request app.
```javascript
const app = require('./app');
const response = await request(app)
```
I got it working by using the correct port, and requiring the app inside the test file.
```javascript
const { PORT } = require('../../config');
require('../../index');
```
```javascript
await page.goto(`http://localhost:${PORT}/apiDocs/`);
await page.goto(`http://localhost:${PORT}/apiDocs/v1.0.0`);
```

**Comment by ewenjo** [comment]
This is my take on it, only problem is that some async operation doesn't stop. No idea why. (Will still work though since tests are run with --detectOpenHandles and --forceExit)
```javascript
const puppeteer = require('puppeteer');

const { PORT } = require('../../config');
const app = require('../../app').listen(PORT);

let browser;

beforeAll(async () => {
  browser = await puppeteer.launch();
});

afterAll(async () => {
  await browser.close();
  app.close();
});

describe('GET /apiDocs/', () => {
  it('Should display "Available versions" text on page', async done => {
    const url = `http://localhost:${PORT}/apiDocs/`;
    const page = await browser.newPage();
    await page.goto(url);
    await expect(page).toMatch('Available versions');
    done();
  });
});

describe('GET /apiDocs/v1.0.0', () => {
  it('Should display "CodingGarden Community App APIs" text on page', async done => {
    const url = `http://localhost:${PORT}/apiDocs/v1.0.0`;
    const page = await browser.newPage();
    await page.goto(url);
    await expect(page).toMatch('CodingGarden Community App APIs');
    done();
  });
});
```

**Comment by tejas-hosamani** [comment]
You were right. That worked!

File: src/api/docs/doc.test.js
Comment (ewenjo) [inline src/api/docs/doc.test.js]: Remove these lines:

```javascript
require('dotenv').config();
require('../../app');
```

File: src/api/docs/doc.test.js
Comment (ewenjo) [inline src/api/docs/doc.test.js]: Replace this line with:
```javascript
require('../../app').listen(PORT);
```
Avoids unnecessary console.log

**Comment by spiray** [comment]
@ewenjo & @CodingCreate101 When I run these tests locally I am getting timeouts. Jest by default gives 5,000 ms for tests to run but you can add a 3rd arg to the `it` call with a longer timeout. Is anyone else running into this issue? 
There is no need to `await` the expect assertions in the test as expect and its methods don't return AFAIK.

I don't know if I am missing something here.

**Comment by ewenjo** [comment]
You're right about the await thing, it should be removed. I'm not getting any timout errors, not sure what the problem could be,

**Comment by spiray** [comment]
Never mind they are passing now.

File: src/api/docs/doc.routes.js
Comment (spiray) [inline src/api/docs/doc.routes.js]: You can destructure this assignment to the 2 methods that are used.

File: src/api/docs/doc.test.js
Comment (spiray) [inline src/api/docs/doc.test.js:22]: This await should be removed.

File: src/api/docs/doc.test.js
Comment (spiray) [inline src/api/docs/doc.test.js:29]: This await should be removed.

File: src/routes.js
Comment (spiray) [inline src/routes.js]: Can you change this variable name to `doc` or `docs` to keep it more consistent?

**Review by spiray — CHANGES_REQUESTED** [review]

**Comment by spiray** [comment]
@CodingCreate101 See my last few comments above. Once those are fixed we should be good to merge! Thanks for your time and contributions :)

File: src/api/docs/doc.test.js
Comment (ewenjo) [reply src/api/docs/doc.test.js:29]: Wait, it should be there. Tried to remove it and tests failed. Some sort of async operation when converting the page to its content.

**Comment by tejas-hosamani** [comment]
Tests are running okay for me; even with `headless:false, slowmo:100`.
Tests are failing for me when I remove `await` for `expect` with `Error: Navigation failed because browser has disconnected!`.
I am thinking, maybe, the browser closes before `page.goto` has any chance to run.

**Comment by ewenjo** [comment]
@CodingCreate101 Take a look at my comment above, the await should be there.

File: src/api/docs/doc.test.js
Comment (spiray) [reply src/api/docs/doc.test.js:29]: My bad!  You are right.

File: src/api/docs/doc.test.js
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js:22]: Maybe, the browser closes before `page.goto` method has any chance to run. Tests are failing for that.

File: src/routes.js
Comment (tejas-hosamani) [reply src/routes.js]: `documentation`? To keep it descriptive? Like all other names.. like video, history, etc.

Comment (ewenjo) [reply src/routes.js]: Documentation works

**Comment by tejas-hosamani** [comment]
Thanks! 👯‍♂  My pleasure. 💯 :) And thank you both for reviews. Learned a lot!

**Review by spiray — APPROVED** [review]

**Pull request merged**
